apps/wallet/src/ui/app/components/transactions-card/Transaction.tsx

```diff
@@ -0,0 +1,205 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import {
+    getExecutionStatusType,
+    getTransactionKindName,
+    getMoveCallTransaction,
+    getExecutionStatusError,
+    getTransferObjectTransaction,
+} from '@mysten/sui.js';
+import { useMemo } from 'react';
+import { Link } from 'react-router-dom';
+
+import { TxnTypeLabel } from './TxnActionLabel';
+import { TxnIcon } from './TxnIcon';
+import { TxnImage } from './TxnImage';
+import { CoinBalance } from '_app/shared/coin-balance';
+import { Text } from '_app/shared/text';
+import { getEventsSummary, getAmount, formatDate } from '_helpers';
+import { useMiddleEllipsis } from '_hooks';
+import { getTxnEffectsEventID } from '_redux/slices/txresults';
+import Alert from '_src/ui/app/components/alert';
+
+import type { SuiTransactionResponse, SuiAddress } from '@mysten/sui.js';
+
+const TRUNCATE_MAX_LENGTH = 8;
+const TRUNCATE_PREFIX_LENGTH = 4;
+
+export function Transaction({
+    txn,
+    address,
+}: {
+    txn: SuiTransactionResponse;
+    address: SuiAddress;
+}) {
+    const { certificate } = txn;
+    const executionStatus = getExecutionStatusType(txn);
+    const txnKind = getTransactionKindName(certificate.data.transactions[0]);
+    const { coins: eventsSummary } = getEventsSummary(txn.effects, address);
+
+    const objectId = useMemo(() => {
+        const transferId = getTransferObjectTransaction(
+            certificate.data.transactions[0]
+        )?.objectRef?.objectId;
+        return transferId
+            ? transferId
+            : getTxnEffectsEventID(txn.effects, address)[0];
+    }, [address, certificate.data.transactions, txn.effects]);
+
+    const amountByRecipient = getAmount(
+        certificate.data.transactions[0],
+        txn.effects
+    );
+
+    const amount = useMemo(() => {
+        const amount = amountByRecipient && amountByRecipient?.[0]?.amount;
+        const amountTransfers = eventsSummary.reduce(
+            (acc, { amount }) => acc + amount,
+            0
+        );
+
+        return Math.abs(amount || amountTransfers);
+    }, [amountByRecipient, eventsSummary]);
+
+    const recipientAddress = useMemo(() => {
```

Comment:
I think ideally we'd want to move all this derivation logic out of the component into re-reusable hooks that could be composed into something for this view. especially if we're going to re-use this logic in other parts of the app, especially if we're going to need to use this logic elsewhere

Reply:
I am not sure where else we will use this